### PR TITLE
Fix SnapshotBasedRecoveryIT#testSnapshotBasedRecovery

### LINKED ISSUE #76595: [CI] SnapshotBasedRecoveryIT testSnapshotBasedRecovery failing
Normally I'd skip timeout errors because they are often just due to luck. But this test is new and already failed 38 times in past few days.

**Build scan**:
https://gradle-enterprise.elastic.co/s/523u2dz3olx3u

**Repro line**:
./gradlew ':qa:rolling-upgrade:v7.14.1#oneThirdUpgradedTest' -Dtests.class="org.elasticsearch.upgrades.SnapshotBasedRecoveryIT" -Dtests.method="testSnapshotBasedRecovery" -Dtests.seed=15D06BC7AFD0B42F -Dtests.bwc=true -Dtests.locale=ca-ES -Dtests.timezone=Africa/Banjul -Druntime.java=8

**Reproduces locally?**:
No

**Applicable branches**:
7.x

**Failure history**:
https://gradle-enterprise.elastic.co/scans/tests?search.relativeStartTime=P7D&search.timeZoneId=Australia/Melbourne&tests.container=org.elasticsearch.upgrades.SnapshotBasedRecoveryIT&tests.sortField=FAILED&tests.test=testSnapshotBasedRecovery&tests.unstableOnly=true

**Failure excerpt**:
```


java.lang.AssertionError: timed out waiting for green state for index [snapshot_based_recovery] cluster state [{ |  
-- | --
  | "cluster_name" : "v7.14.1", |  
  | "cluster_uuid" : "vNuyvMUdTfaVJAKtMymzNQ", |  
  | "version" : 620, |  
  | "state_uuid" : "N5K_VdvhSVupuKwpFqXHBw", |  
  | "master_node" : "2YFxMyn6Rrmw9JqR-zgMRg", |  
  | "blocks" : { |  
  | "indices" : { |  
  | "index_mixed_7140199" : { |  
  | "4" : { |  
  | "description" : "index closed", |  
  | "retryable" : false, |  
  | "levels" : [ |  
  | "read", |  
  | "write" |  
  | ] |  
  | } |  
  | }, |  
  | "closed_index_replica_allocation" : { |  
  | "4" : { |  
  | "description" : "index closed", |  
  | "retryable" : false,


```

### PR DESCRIPTION
**Pull request by fcofdez**
Move the shard to a replica in an older version when the primary
is located in the upgraded node during the first rolling restart
round.

Closes #76595

### CODE REVIEW

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java [hunk 1]
```diff
@@ -65,16 +73,38 @@ public void testSnapshotBasedRecovery() throws Exception {
...
+                            Settings.builder()
+                                .put("index.routing.allocation.exclude._id", primaryNodeId)
+                                .build()
+                        );
+                        cancelShard(indexName, 0, primaryNodeId);
+
+                        String currentPrimaryNodeId = getPrimaryNodeIdOfShard(indexName, 0);
+                        assertThat(getNodeVersion(currentPrimaryNodeId), is(equalTo(UPGRADE_FROM_VERSION)));
+
+                        updateIndexSettings(
```
Comment (henningandersen) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java:97]: I think this risks balancing back the shard to the upgraded node. You can disable balancing and reenable it after reestablishing replica. Or we could just keep the exclusion until the next round. I think I prefer the latter.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java [hunk 2]
```diff
@@ -65,16 +73,38 @@ public void testSnapshotBasedRecovery() throws Exception {
...
-                // on the upgraded node. This causes issues when removing and adding replicas, since then we cannot allocate to
-                // any of the old nodes. That is an issue only for the first mixed round, hence this check.
-                // Ideally we would find the reason the primary ends on the upgraded node and fix that (or figure out that it
-                // is all good).
-                // @AwaitsFix(bugUrl = https://github.com/elastic/elasticsearch/issues/76595)
-                if (CLUSTER_TYPE != MIXED || FIRST_MIXED_ROUND == false) {
-                    // Drop replicas
-                    updateIndexSettings(indexName, Settings.builder().put(IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0));
+                if (FIRST_MIXED_ROUND) {
+                    String primaryNodeId = getPrimaryNodeIdOfShard(indexName, 0);
```
Comment (henningandersen) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java:77]: I wonder if we should not instead just find the node on the new version and put the exclusion in always? I think there is a risk of the shard moving between the check here and adding replicas below.

**Review by henningandersen — COMMENTED** [review]
I left a couple of comments.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java [hunk 3]
```diff
@@ -65,16 +74,43 @@ public void testSnapshotBasedRecovery() throws Exception {
...
+                        indexName,
+                        Settings.builder()
+                            .putNull("index.routing.allocation.exclude._id")
+                            .build()
+                    );
                 }
+
+                // Drop replicas
+                updateIndexSettings(indexName, Settings.builder().put(IndexMetadata.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 0));
                 ensureGreen(indexName);
```
Comment (henningandersen) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java:114]: I do not think this `ensureGreen` does anything?

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java [hunk 4]
```diff
@@ -87,6 +123,97 @@ public void testSnapshotBasedRecovery() throws Exception {
         }
     }
 
+    private void updateIndexSettings(String indexName, Settings settings) throws IOException {
```
Comment (henningandersen) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/SnapshotBasedRecoveryIT.java:126]: This looks nearly identical to the static `updateIndexSettings` inherited from `ESRestTestCase`, do we need this?

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Pull request merged**